### ISSUE TITLE
[Truffle] Rubinius::Type.check_convert_type

### ISSUE THREAD

**Issue by bjfish** [jruby/jruby#3992]
This `Rubinius::Type.check_convert_type` comment says that it corresponds to `rb_check_convert_type` but it appears that it doesn't handle `respond_to_missing?` or `missing_method` methods.

Correcting this should fix the `array/flatten_spec.rb` `flatten performs respond_to? and method_missing-aware checks when coercing elements to array`.

`Rubinius::Type.coerce_to` should probably be reviewed too.

**Comment by eregon** [comment]
This should be handled by the default `respond_to?`, no?
`check_convert_type` uses `vm_object_respond_to`, which uses `Kernel#respond_to?`, so I am not sure what's going wrong here.

**Comment by eregon** [comment]
The issue seems to be there is no equivalent of `rb_check_funcall` at the Ruby level.

**Comment by bjfish** [comment]
@eregon Could there be an issue with `vm_object_respond_to` ? The following example:

```
class Test
  def respond_to?(method_name, include_private = false)
    puts "IN RESPOND TO"
    true
  end
end
test = Test.new
puts "test.respond_to? :to_ary - #{test.respond_to? :to_ary}"
puts "Rubinius::Type.object_respond_to? test, :to_ary, true - #{Rubinius::Type.object_respond_to? test, :to_ary, true}"
```
Outputs:
```
IN RESPOND TO
test.respond_to? :to_ary - true
Rubinius::Type.object_respond_to? test, :to_ary, true - false
```

**Comment by eregon** [comment]
@bjfish Yes, `vm_object_respond_to` is bound to `BasicObject#respond_to?`, so indeed we should probably try calling respond_to? dynamically. I tried a few things but gave up since it seems there are many differences here with MRI and it's quite hard to follow the code.

**Comment by bjfish** [comment]
This was fixed by #4146
